Added support for unencrypted values and branches

twolfson opened the pull request:
In #24, we discussed potential solutions for unencrypted values and branches. This PR implements the agreed upon feature (using a key suffix to denote an unencrypted value/branch). In this PR:

- Added tests for encryption and decryption
    - We paid attention to make sure the MAC included unencrypted values to prevent their tampering
- Implemented unencrypted values/branches support on encryption/decryption
- Added CLI support for `--unencrypted-suffix`
- Added unencrypted suffix persistence by storing it under `tree['sops']['unencrypted_suffx']`
- Fixes #24 

**Notes:**

We did our best to keep code style consistent but due to the functional design we wound up passing around `unencrypted`/`unencrypted_suffx` around a lot =(

We weren't sure about when to camelCase kwargs vs snake_case kwargs so we went with the Python convention of snake_case.

twolfson commented on sops/__init__.py:
```diff
@@ -202,10 +207,17 @@ def main():
+    if args.unencrypted_suffix:
+        unencrypted_suffix = args.unencrypted_suffix
+    elif 'sops' in tree and 'unencrypted_suffix' in tree['sops']:
+        unencrypted_suffix = tree['sops']['unencrypted_suffix']
```
I'm not fully certain if this is the place we want to load this value but I'm not sure of any place better

jvehent replied:
I might fit better into `initialize_tree()` and modify the `UNENCRYPTED_SUFFIX` there.

twolfson replied:
Ah, alright. I will try that out

twolfson commented:
It looks like Travis CI caught some regressions about value types (e.g. `b'a'` vs `'a'`). I'm not sure what direction we want to take (e.g. still use `iv`, `tag,` and `type` on unencrypted values OR determine if the sounds of this is due to not using something like `six`) so I will wait for feedback.

jvehent commented on sops/__init__.py:
```diff
@@ -95,6 +95,8 @@
 
 INPUT_VERSION = VERSION
 
+DEFAULT_UNENCRYPTED_SUFFIX = '_unencrypted'
```
What about just `UNENCRYPTED_SUFFIX` ? We might modify it in `initialize_tree`, so it may not always be a default.

twolfson replied:
But this is a `DEFAULT_` in the sense of `DEFAULT_TEXT` and `DEFAULT_JSON`, isn't it?

jvehent replied:
Up to you how you want to handle it. You could keep the default global, and create another global to store the current value.

twolfson replied:
I think based on how we handle `INPUT_VERSION` and our delay for #3, I will likely move to a global `UNENCRYPTED_SUFFIX` for developer sanity. I might keep a `DEFAULT_` still since it's nice to declare constants too.

jvehent commented on sops/__init__.py:
```diff
@@ -820,6 +840,10 @@ def encrypt(value, key, aad=b'', stash=None, digest=None):
+    # if we don't want to encrypt, then return the value
+    if unencrypted:
+        return value
+
```
If we know in advance we don't want to encrypt the value, it would be preferable to just not enter `encrypt()`

twolfson replied:
This was placed in here to handle `digest.update` since it should still be part of the MAC. Did we want to relocate `digest` computation outside of `encrypt` altogether?

We could also consider resolving this by renaming `encrypt` to something like `serialize` or `dumps`

jvehent replied:
I think a cleaner approach would be to do a first pass on the tree to calculate the digest, then do a second pass to encrypt it. It's beyond the scope of this PR though, so leave it as it is.

twolfson replied:
Okie doke

jvehent commented on sops/__init__.py:
```diff
@@ -202,10 +207,17 @@ def main():
         # Encrypt mode: encrypt, display and exit
         key, tree = get_key(tree, need_key)
-        tree = walk_and_encrypt(tree, key)
+        tree = walk_and_encrypt(tree, key, unencrypted_suffix=unencrypted_suffix)
```
Since we can only have one `UNENCRYPTED_SUFFIX` per invocation of sops, I think it would make the code more readable to just rely on a global variable, instead of passing the value around.

twolfson replied:
That feels pretty dirty, I mentioned this in the initial PR notes but maybe a classical structure for trees would be better (e.g. save `unencrypted_suffix`, `digest`, and `iv` to `self` instead of passing them around)?

jvehent replied:
Agreed. To be addressed in #3 then.

twolfson replied:
:+1:

twolfson replied:
I will explore shifting to a global variable

jvehent commented on tests/test_sops.py:
```diff
@@ -175,6 +211,26 @@ def test_bytes_encrypt_and_decrypt(self):
+            tree = sops.load_file_into_tree('path', 'json')
+        tree['sops'] = dict()
+        crypttree = sops.walk_and_encrypt(OrderedDict(tree), key, isRoot=True)
+        assert tree['sops']['mac'].startswith("ENC[AES256_GCM,data:")
```
would be good to assert that the unencrypted values are still in cleartext here.

twolfson replied:
The encrypt unencrypted test covers that test case already =/

jvehent commented on sops/__init__.py:
```diff
@@ -620,6 +632,7 @@ def walk_and_decrypt(branch, key, aad=b'', stash=None, digest=None,
     for k, v in branch.items():
         if k == 'sops' and isRoot:
             continue    # everything under the `sops` key stays in clear
+        v_unencrypted = unencrypted or k.endswith(unencrypted_suffix)
```
what about `must_decrypt = not unencrypted and not k.endswith(UNENCRYPTED_SUFFIX)` ?
and then only:
```python
if must_decrypt:
    decrypt(....)
```

twolfson replied:
We used the same language (i.e. `unencrypted`) for consistency to make the connection obvious to programmers.

Additionally, we need to pass `v_unencrypted` (or whatever we call it) to nested branch traversals since they need to not encrypt themselves as well (e.g. `{'a_unencrypted': {'b': 'c'}}`). As a result, performing boolean flipping will probably cause more confusion than help =/

twolfson replied:
This comment has been resolved further down. We will rename it to `unencrypted_branch` for clarity

jvehent commented on sops/__init__.py:
```diff
@@ -670,17 +683,22 @@ def walk_list_and_decrypt(branch, key, aad=b'', stash=None, digest=None):
+        if digest:
+            digest.update(value)
+        return value
+
```
humf.... looks like tangling the digest calculation with `decrypt()` wasn't a great idea after all...

jvehent commented on sops/__init__.py:
```diff
@@ -740,13 +758,14 @@ def decrypt(value, key, aad=b'', stash=None, digest=None):
     for k, v in branch.items():
         if k == 'sops' and isRoot:
             continue    # everything under the `sops` key stays in clear
+        v_unencrypted = unencrypted or k.endswith(unencrypted_suffix)
```
same comment as with decrypt (I find `v_unencrypted` to lack expliciteness).

twolfson replied:
We can rename this to `unencrypted_branch` if that helps with clarity

jvehent replied:
`unencrypted_branch` does seem a bit clearer

jvehent commented on sops/__init__.py:
```diff
@@ -754,17 +773,18 @@ def walk_and_encrypt(branch, key, aad=b'', stash=None,
             branch[k] = ruamel.yaml.scalarstring.PreservedScalarString(ev)
         else:
-            branch[k] = encrypt(v, key, aad=caad, stash=nstash, digest=digest)
+            branch[k] = encrypt(v, key, aad=caad, stash=nstash, digest=digest, unencrypted=v_unencrypted)
```
here the digest calculation doesn't depend on entering `encrypt()`, so let's not enter it at all.

twolfson replied:
I don't follow, can you explain further?

jvehent replied:
Nevermind. For some reason, I misread my own code. This is fine as it is.

jvehent commented:
> It looks like Travis CI caught some regressions about value types (e.g. b'a' vs 'a').
> I'm not sure what direction we want to take (e.g. still use iv, tag, and type on
> unencrypted values OR determine if the sounds of this is due to not using something
> like six) so I will wait for feedback.

This is happening because you enter `encrypt()` even for unencrypted values, and only return the original value after type detection and conversation happens. `encrypt` converts the value to `bytes`, which changes the `JSON` string you're expecting.
```python
def encrypt(value, key, aad=b'', stash=None, digest=None):
.......
     if not isinstance(value, bytes):
         # if not bytes, convert to bytes
         value = str(value).encode('utf-8')
```
If you don't enter `encrypt()` at all for unencrypted values, or return prior to the type convertion, the problem should go away.

jvehent commented:
This is great work! Thanks a lot for taking on that issue.  The code looks good overall, and modulo addressing the review comments, I think it can be merged fairly rapidly. +:100:

twolfson commented:
Thanks for the quick and thorough review =) I left some counter comments. I also alluded to a classical restructure of `sops`. I will open a new issue to have that discussion with pros/cons listed.

twolfson commented:
To be clear, I'm waiting on responses to the comments I left before iterating on the PR again (so all updates can be done in 1 fell swoop)

jvehent commented:
Ok, we have a chicken and egg problem: If a value is unencrypted, you want to exit `encrypt()` before it gets converted to bytes. But you still want it added to the digest, which requires converting it to bytes.
`encrypt()` needs to be rewritten like this:
```python
     if not isinstance(value, bytes):
         # if not bytes, convert to bytes
         bvalue = str(value).encode('utf-8')
 
     if digest:
         digest.update(bvalue)

    if unencrypted:
        return value

    value = bvalue
```

twolfson commented:
Cool, I think I'm good to revisit this again. I will take another attempt by the end of the weekend.

twolfson commented:
This PR has been updated with the following changes:

- Renamed `v_unencrypted` to `unencrypted_branch`
- Moved to use `DEFAULT_UNENCRYPTED_SUFFIX` and `UNENCRYPTED_SUFFIX` globals
- Relocated `sops['unencrypted_suffix']` setter to emulate behavior for `INPUT_VERSION`
    - Unfortunately, this means we had to drop an `assert tree['sops']['unencrypted_suffix']` from our tests since there are no tests for `update_master_keys` (and adding one would be verbose/out of scope) =(
- Repaired broken Python 3.4 encryption
    - We deviated a bit from the suggestions; we decided to always perform `if unencrypted: if digest` checks at the head of each `encrypt`/`decrypt` function. This will prevent accidental regressions from developers missing the `unencrypted` edge case
    - We added a `to_bytes` function to simplify reused encoding in `encrypt`/`decrypt` for `unencrypted` content

jvehent commented:
Code looks good and tests pass, this is good to merge!
I'll test a bit locally and push a couple more unrelated change before publishing 1.7 on pypi.

Thanks a lot for the contribution @twolfson ! +:100:

twolfson commented:
Awesome, thanks =D
